### PR TITLE
Introduce SkyWalking v9 core query protocol.

### PR DESCRIPTION
**Pull request by wu-sheng**
I have added descriptions in the comments, please take a look.

### CODE REVIEW

File: metadata-v2.graphqls [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+extend type Query {
+    # Read the service list according to layer.
+    listServices(layer: String!): [Service!]!
+    # Read service instance list.
+    listInstances(duration: Duration!, serviceId: ID!): [ServiceInstance!]!
+
+    # Search and find service according to given ID. Return null if not existing.
+    findService(serviceId: String!): Service
+    # Search and find service instance according to given ID. Return null if not existing.
+    findInstance(endpointId: String!): ServiceInstance
```
Comment (Fine0830) [inline metadata-v2.graphqls:100]: `endpointId` should be `instanceId` ?

Comment (wu-sheng) [reply metadata-v2.graphqls:100]: Yes

**Comment by Fine0830** [comment]
One question: can we add `listLayers` for infrastructure map.

**Comment by wu-sheng** [comment]
I think we don't have layer on the top of layer for now.

Current infrastructure map should be inside one layer. Group -> Service -> Instance. Is this working for you?

**Comment by Fine0830** [comment]
> 

Yes, it's fine. I forgot we added sub-menus for infrastructure.

**Review by Fine0830 — APPROVED** [review]

**Review by wankai123 — APPROVED** [review]

**Review by kezhenxu94 — COMMENTED** [review]
Hi sorry for late, just one question: is this v9 query protocol compatible with previous one? It looks that you want to be compatible as much as possible but there are some deleted fields, which should be breaking changes I think?

**Comment by wu-sheng** [comment]
> Hi sorry for late, just one question: is this v9 query protocol compatible with previous one? It looks that you want to be compatible as much as possible but there are some deleted fields, which should be breaking changes I think?

There are just grammar level compatible, and provide limited features. Such as grouping based query is replaced by layer based query.
Could you share which specific one do you have concern?

File: topology.graphqls [hunk 2]
```diff
@@ -58,12 +63,11 @@ type ServiceInstanceNode {
...
-    # The type name may be
-    # 1. The service provider/middleware tech, such as: Tomcat, SpringMVC
-    # 2. Conjectural Service, e.g. MySQL, Redis, Kafka
+
+    # Deprecated since 9.0.0. Instance is an executable unit of service, it doesn't have its own type.
     type: String
-    # It is a conjecture node or real node, to represent an instance.
-    isReal: Boolean!
+    # Deprecated since 9.0.0. Don't need to use this anymore
+    isReal: Boolean
```
Comment (kezhenxu94) [inline topology.graphqls:70]: > Could you share which specific one do you have concern?

Not a critical concern but looking this kind of changes reminded me that if the new backend returns `null` for this field, the previous consumer of this protocol will break, but if the backend keeps returning non-null value then adding the deprecate message should be enough.

Comment (wu-sheng) [reply topology.graphqls:70]: I think we should return empty string, even the field is changed into `nullable`.

Comment (kezhenxu94) [reply topology.graphqls:70]: > I think we should return empty string, even the field is changed into `nullable`.

OK, that makes sense to me

Comment (wu-sheng) [reply topology.graphqls:70]: And, the key idea to keep this compatibility to make e2e, CLI, and integration(like HPA) work as usual. But the rocketbot UI is not considered that also working as usual.

**Pull request merged**